### ISSUE TITLE
Same job running multiple times

### ISSUE THREAD

**Issue by ernestorocha** [vsivsi/meteor-job-collection#99]
I still need to investigate better, but for some reason my jobs are running multiple times. I'm on `1.1.4`, and this issues only happens if I update to `1.2.x`.

My console looks like that:

```shell
W20150806-10:41:19.143(-4)? (STDERR) Running job not found NGkyq3yFqQDcm9LKZ vKpn4QjD6AaokTjMQ
W20150806-10:41:19.485(-4)? (STDERR) Running job not found NGkyq3yFqQDcm9LKZ vKpn4QjD6AaokTjMQ
W20150806-10:41:21.471(-4)? (STDERR) jobFail failed
W20150806-10:41:21.595(-4)? (STDERR) Running job not found ZWdb2toxbiruqy3DF eWsGSmDWMFhjNWFFL
W20150806-10:41:23.874(-4)? (STDERR) Running job not found NGkyq3yFqQDcm9LKZ vKpn4QjD6AaokTjMQ
```
or
```shell
W20150806-10:40:58.095(-4)? (STDERR) Running job not found p6sJeR5qTWKeXAZmp SbJriNhSE8txN9ZiG
W20150806-10:40:58.185(-4)? (STDERR) jobFail failed
W20150806-10:40:58.187(-4)? (STDERR) jobFail failed
W20150806-10:40:58.299(-4)? (STDERR) Running job not found p6sJeR5qTWKeXAZmp SbJriNhSE8txN9ZiG
W20150806-10:40:59.554(-4)? (STDERR) jobFail failed
W20150806-10:40:59.562(-4)? (STDERR) jobFail failed
W20150806-10:40:59.931(-4)? (STDERR) jobFail failed
```
Any idea ?

**Comment by vsivsi** [comment]
Hi, is this consistently reproducible?  Can you share some simple code that causes it to happen?

Beyond that, I'll need to know a lot more about your app, the environment it runs in, etc. in order to help. The above console outputs don't reveal much.  At a minimum you should try turning on the full console debug output so that you can see all of the Method calls that lead to these warnings.

```js
// Log everything to stdout
jc.setLogStream(process.stdout);
```

**Comment by ernestorocha** [comment]
I'm going to investigate this right now, but for instance this is my code, for example, sometimes more than 2 emails are sent:

```javascript
emailJobs = JobCollection('emailNotifications');

var emailNotifcationWorker = emailJobs.processJobs('sendNotification', {workTimeout: 3*60*1000}, function(job, cb) {
   var data = job.data;

   sendPrettyEmailNotification(data.nId);

   cb();
 });
```

And then later on the code: 
```javascript
 Notifications.sendEmailNotification = function(nId) {
   var n = Notifications.findOne(nId);
   var user = n.targetUser();

   if(_.contains(user.notificationEvents(), n.event)) {
     var job = new Job('emailNotifications', 'sendNotification', {nId: nId});

     job.priority('high').retry({
       retries: 5,
       wait: 1*6*1000
     }).delay(1).save();
   }
 };
```
I don't think it's this hook since only happens when I upgrade jobcollection package.
```javascript
Notifications.after.insert(function(userId, doc) {

   Notifications.sendEmailNotification(doc._id);

 });
```

But since I know this package has some problems, I'll try to eliminate all possibilities.

Thanks!

**Comment by vsivsi** [comment]
At first glance it looks like your worker callback function to `processJobs` sends the email, but never calls `job.done()` to alert the queer that the work was completed successfully.  Because you set a worker timeout, that job will then auto-fail on the server and be reassigned to a new worker, that will send the email, and not call `job.done()`, etc.

**Comment by ernestorocha** [comment]
It seems that this it not related to timeout optitons, but to the "retry" since it's set only to 6 seconds in this case.

**Comment by vsivsi** [comment]
That's correct.  `.retry(2)` means "try it once, and if that fails retry two more times" for a total of three attempts. 

I believe the root cause of your issue is as I said above, you need to call `job.done()` in your worker function.

**Comment by ernestorocha** [comment]
Sorry, I misunderstood what you said. So, it's my responsibility  to mark the job as completed ?

If so, I'll just add that line on my code then report back here! Thanks!

**Comment by vsivsi** [comment]
Right. The worker needs to eventually call either `job.done()` or `job.fail()`.  Because job-collection has no insight into the workload, you must tell it whether a given attempt was successfully completed or not via one of those two calls. If you don't then if you use the `workerTimeout` option, the server will eventually "auto-fail" the job so it can be retried later.

**Comment by a4xrbj1** [comment]
@ernestorocha, please do also check your workTimeout parameter. That was causing me the same trouble right now, even though I used the proper job.done() to finish the job it got re-submitted again as the Zombie cleanup process (steered via this workTimeout parameter) made it fail & re-submit beforehand (before the job.done was executed).
